File: microservices-project/email-service/src/services/email.service.ts
import { resend } from "../config/email";
import { paymentFailedTemplate } from "../templates/paymentFailedTemplate";
import { paymentRefundTemplate } from "../templates/pymentRefundTemplate";
import { paymentSuccessTemplate } from "../templates/paymentSuccessTemplate";
import { newUserTemplate } from "../templates/newUserTemplate";

export enum EmailType {
  PAYMENT_CONFIRMED = "PAYMENT_CONFIRMED",
  PAYMENT_FAILED = "PAYMENT_FAILED",
  REFUND_INITIATED = "REFUND_INITIATED",
  NEW_USER = "new_user",
}

export const templateMap: Record<EmailType, (data: any) => string> = {
  [EmailType.PAYMENT_CONFIRMED]: paymentSuccessTemplate,
  [EmailType.PAYMENT_FAILED]: paymentFailedTemplate,
  [EmailType.REFUND_INITIATED]: paymentRefundTemplate,
  [EmailType.NEW_USER]: newUserTemplate,
};

export const subjectMap: Record<EmailType, string> = {
  [EmailType.PAYMENT_CONFIRMED]: "Payment Successful",
  [EmailType.PAYMENT_FAILED]: "Payment Failed",
  [EmailType.REFUND_INITIATED]: "Refund Initiated",
  [EmailType.NEW_USER]: "Welcome to ...",
};

const isEmailType = (value: string): value is EmailType => {
  return Object.values(EmailType).includes(value as EmailType);
};

export const sendMail = async (email: string, type: string, data: any) => {
  if (!isEmailType(type)) {
    throw new Error(`Invalid email type provided: ${type}`);
  }

  const templateFn = templateMap[type];

  const message = {
    from: "Email service <no-reply@sneax.quest>",
    to: email,
    subject: subjectMap[type],
    html: templateFn(data),
  };

  return await resend.emails.send(message);
};

File: microservices-project/email-service/internal/services/email_service.go
package services

import (
	"fmt"
	"log/slog"

	"github.com/SneaX-23/Microservices-project/microservices-project/email-service/internal/templates"
)

type ServiceConfig struct {
	LoginURL string
}

type EmailService struct {
	mailer   Mailer
	renderer *templates.Renderer
	config   ServiceConfig
}

func NewEmailservice(mailer Mailer, renderer *templates.Renderer, config ServiceConfig) *EmailService {
	return &EmailService{
		mailer:   mailer,
		renderer: renderer,
		config:   config,
	}
}

func (s *EmailService) SendEmail(to, emailType string, data any) error {
	var templateName string
	var subject string

	// Map the messageType(emailType) to a template file
	switch emailType {
	case "PAYMENT_CONFIRMED":
		templateName = "payment_success.html"
		subject = "Payment Successful"
	case "PAYMENT_FAILED":
		templateName = "payment_failed.html"
		subject = "Payment Failed"
	case "new_user":
		templateName = "new_user.html"
		subject = "Welcome!"
	case "REFUND_INITIATED":
		templateName = "payment_refund.html"
		subject = "Refund Initiated"
	default:
		slog.Error("Unknown email type", "type", emailType)
		return fmt.Errorf("unknown email type: %s", emailType)
	}

	// Render the body
	body, err := s.renderer.Render(templateName, data)
	if err != nil {
		return err
	}

	// Send via Mailer
	return s.mailer.Send(to, subject, body)
}
